### PR TITLE
Add "compare" on dataclass fields (Fixes: #8905)

### LINKED ISSUE #8905: Missing support for 'compare' on dataclass field
### Describe the bug

I'm playing around with the annotated declarative setups for my Models and realized that we are missing support for the `compare` argument for `dataclasses.field()`.

Support for the general `eq` parameter (the toggle to build `__eq__` on dataclass generation) on class-level is available, what I'm missing is the corresponding `compare` boolean-argument on field-level that defines if a field should be used for comparison by `__eq__`.

It would be really helpful to have that one supported, since we often have some kind of dynamic data on initialization of database models ('created/changed timestamp', 'UUID as primary key', ...) that might not be desired/relevant for comparison.

Proposal for adding support for the compare argument attached as PR.

### To Reproduce

```python
import uuid

class A(dc_decl_base):
    __tablename__ = "a"

    id: Mapped[uuid.UUID] = mapped_column(
        primary_key=True,
        default_factory=uuid.uuid4,
        init=False,
        compare=False)
    data: Mapped[str]


a1 = A(data='foo')
a2 = A(data='foo')
assert a1 == a2
```

### Error

```
Traceback (most recent call last):
  File "/home/simon/repos/thirdparty/sqlalchemy/test/orm/declarative/test_dc_transforms.py", line 570, in test_compare
    class A(dc_decl_base):
  File "/home/simon/repos/thirdparty/sqlalchemy/test/orm/declarative/test_dc_transforms.py", line 573, in A
    id: Mapped[int] = mapped_column(primary_key=True, compare=False)
  File "/home/simon/repos/thirdparty/sqlalchemy/.tox/py/lib/python3.10/site-packages/sqlalchemy/orm/_orm_constructors.py", line 256, in mapped_column
    return MappedColumn(
  File "/home/simon/repos/thirdparty/sqlalchemy/.tox/py/lib/python3.10/site-packages/sqlalchemy/orm/properties.py", line 577, in __init__
    self.column = cast("Column[_T]", Column(*arg, **kw))
  File "/home/simon/repos/thirdparty/sqlalchemy/.tox/py/lib/python3.10/site-packages/sqlalchemy/sql/schema.py", line 1977, in __init__
    self._extra_kwargs(**dialect_kwargs)
  File "/home/simon/repos/thirdparty/sqlalchemy/.tox/py/lib/python3.10/site-packages/sqlalchemy/sql/schema.py", line 2038, in _extra_kwargs
    self._validate_dialect_kwargs(kwargs)
  File "/home/simon/repos/thirdparty/sqlalchemy/.tox/py/lib/python3.10/site-packages/sqlalchemy/sql/base.py", line 547, in _validate_dialect_kwargs
    raise TypeError(
TypeError: Additional arguments should be named <dialectname>_<argument>, got 'compare'

``` 

### Versions

- OS: Linux (Debian, testing)
- Python: 3.10
- SQLAlchemy: 2.0.0b4
- Database: sqlite (memory)
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): -

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by simonschiele**
Add support for `compare` argument on dataclass field. This allows to conveniently remove single fields from getting used by dataclasses `__eq__` comparison.

### Description
Within annotated declarative setups for Models currently the `compare` argument is not supported for fields. Would be strongly appreciated to support this one, since it is a very convenient way to control the behavior of the generated `__eq__` for transformed dataclasses.

Change simply adds the parameter to the `_AttributeOptions` used for dataclass-generation, which allows for usage within `mapped_column()` and `relationship()` like for the other supported arguments.

Fixes #8905

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [X] A short code fix
	- fixes 
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 3a9feb5dfc326705998fbba7f389ff08da62520a of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 3a9feb5dfc326705998fbba7f389ff08da62520a: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4254

File: test/orm/declarative/test_dc_transforms.py [hunk 1]
```diff
@@ -565,6 +565,18 @@ class Engineer(MappedAsDataclass, Person):
             ),
         )
 
+    @testing.only_if(lambda: compat.py310, "python 3.10 is required")
```
Comment (CaselIT) [inline test/orm/declarative/test_dc_transforms.py:568]: I think we can remove this, since compare has always been an option of dataclass

**Review by CaselIT — COMMENTED** [review]
Looks ok, just a minor change to the tests

**Comment by simonschiele** [comment]
> hm I wonder why i didnt include "compare" in the first place?

I blame the naming. Due to the feature being controlled on class-level by `eq` and on field level by `compare` I also found this to be highly confusing.

In theory we are also missing `hash` and `metadata` on the field-level. Didn't add these to keep the scope reasonable. Also couldn't come up with a use-case for `hash` on the fly.

File: test/orm/declarative/test_dc_transforms.py (hunk 1 above)
Comment (simonschiele) [reply test/orm/declarative/test_dc_transforms.py]: Changed, thanks.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision ea9a53d2ca60befdd0c570013c0e57a78c11dd4a of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset ea9a53d2ca60befdd0c570013c0e57a78c11dd4a added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4254

**Comment by simonschiele** [comment]
> **mike bayer** (zzzeek) wrote:
> I just pushed [1057b47](https://github.com/sqlalchemy/sqlalchemy/commit/1057b47bca2522e45d9621a709d033aa4fb88888) which means you will need to rebase and add to associationproxy.py also

Thanks :-)

Rebased, adjusted and also added the parameter to the corresponding docstrings.

**Comment by zzzeek** [comment]
oh thanks!  I just rebased on this end and added a few docstirngs, so we are out of sync by one commit.  It's all set, so thanks very much!!

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4254 has been **merged**. Congratulations! :)

**Pull request closed without merging**